src/Servers/IIS/IIS/src/StartupHook.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Security.Claims;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting.Views;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.IIS;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.StackTrace.Sources;
+
+internal class StartupHook
+{
+    public static void Initialize()
+    {
+        // TODO make this unhandled exception
+        AppDomain.CurrentDomain.FirstChanceException += (sender, eventArgs) =>
+        {
+            var exception = eventArgs.Exception;
+
+            var iisConfigData = NativeMethods.HttpGetApplicationProperties();
+            var contentRoot = iisConfigData.pwzFullApplicationPath.TrimEnd(Path.DirectorySeparatorChar);
+
+            var model = new ErrorPageModel
+            {
+                RuntimeDisplayName = RuntimeInformation.FrameworkDescription
+            };
+            var systemRuntimeAssembly = typeof(System.ComponentModel.DefaultValueAttribute).GetTypeInfo().Assembly;
+            var assemblyVersion = new AssemblyName(systemRuntimeAssembly.FullName).Version.ToString();
+            var clrVersion = assemblyVersion;
+            model.RuntimeArchitecture = RuntimeInformation.ProcessArchitecture.ToString();
+            var currentAssembly = typeof(ErrorPage).GetTypeInfo().Assembly;
+            model.CurrentAssemblyVesion = currentAssembly
+                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()
+                .InformationalVersion;
+            model.ClrVersion = clrVersion;
+            model.OperatingSystemDescription = RuntimeInformation.OSDescription;
+
+            var exceptionDetailProvider = new ExceptionDetailsProvider(
+                new PhysicalFileProvider(contentRoot),
+                sourceCodeLineCount: 6);
+
+            model.ErrorDetails = exceptionDetailProvider.GetDetails(exception);
+
+            var errorPage = new ErrorPage(model);
+            var context = new IntermediateHttpContext();
```

Comment:
I expect you can get rid of this mock by adding an overload of ExecuteAsync that writes directly to a Stream.

Reply:
Yeah that would probably be cleaner.

Reply:
I still may need to mock the response. By default, it sets Response.ContentType to text/html no matter what. I'm going to leave the mock for now.

Reply:
That line is here (https://github.com/aspnet/AspNetCore/blob/482577287429ca72a1d10fd367fe43556e3621ac/src/Shared/ErrorPage/ErrorPage.cshtml#L18) and would be easy to move out.